### PR TITLE
Load more entries when near the end

### PR DESCRIPTION
**Pull request by fatlotus**
It's a fairly involved update, but I think the result is worth it. Please let me know if anything looks funny.

Note that this exposes some of the problems we have on misses — because they're not parallelized, it will sometimes take almost 3s to render the fragment. There's stuff we can do to improve this, but it seemed out of scope.

Tagging with #11.

### CODE REVIEW

File: caravel/controllers/listings.py [hunk 1]
```diff
@@ -7,12 +7,25 @@
...
+
+    # Parse filtering options from query.
+    query = request.args.get("q", "")
+    offset = int(request.args.get("offset", "0"))
+    if offset < 0:
+        offset = 0
+
+    # Compute the results matching that query.
+    listings = helpers.run_query(query)
+    listings = itertools.islice(listings, offset, offset + 30)
```
Comment (georgeteo) [inline caravel/controllers/listings.py]: Make the offset a multiple of 4?

File: caravel/storage/helpers.py [hunk 2]
```diff
@@ -57,13 +58,18 @@ def run_query(query=""):
     words = words[:5] # TODO: Raise once we know the approximate cost.
 
     # Retrieve the keys for entities that match all terms.
-    shards = [set(fetch_shard(entities.fold_query_term(w))) for w in words]
+    shards = [fetch_shard(entities.fold_query_term(w)) for w in words]
```
Comment (georgeteo) [inline caravel/storage/helpers.py]: Why no need for set?

Comment (fatlotus) [inline caravel/storage/helpers.py]: We need to keep the keys in order of posting date; set uses hashing, which removes order.

Comment (georgeteo) [inline caravel/storage/helpers.py]: Okay.

**Comment by georgeteo** [comment]
Okay. Please settle merge conflicts. Then looks good.

**Pull request merged**